### DISCUSSION TITLE
Reusable workflows UI - structured dependency view

### DISCUSSION THREAD

**Discussion by vaind** [community/community#51496] [Actions]
### Select Topic Area

Product Feedback

### Body

Moving a set of jobs, with dependencies between them, to a reusable workflow makes the UI for that given workflow just list the inner jobs as a flat list. This makes it very hard to see what's actually being run, see the dependencies and resolve issues.

See the following example of the same job, before and after moving part of the jobs to the reusable workflow:
<img width="1265" alt="image" src="https://user-images.githubusercontent.com/6349682/228868633-a29a2424-87a7-4aa7-be5f-967b25694b6a.png">

<img width="119" alt="image" src="https://user-images.githubusercontent.com/6349682/228868499-df85e05c-633f-4308-a76f-7e4c6655e281.png">

**Comment by sgtsquiggs** [comment]
+1 - this was the discussion I was about to start:

Reusable workflows when called do not display dependencies between jobs.

In this example, my `Changes` job generates variables to use in the dependent matrix. In the matrix, I call another workflow which has several jobs with dependencies. Those dependencies are not displayed in any way that is easy to grok.

Ideally the matrix should show the nested workflows with dependencies, like normal workflows have.

![image](https://user-images.githubusercontent.com/3420325/231816640-a84c7f59-7cce-4ce0-945b-04c9e27c09bf.png)
![image](https://user-images.githubusercontent.com/3420325/231816678-f61603f5-6734-41b4-a7cd-414459504cd3.png)

**Comment by cfoch** [comment]
I have the following structure:

scheduler workflow (*) -> matrix generator workflow -> reusable workflow (14 instances)

(*) passes some inputs to the matrix generator workflow to filter the matrix, depending on if it is a PR or Push or wahtever.



![Screenshot from 2023-07-11 07-54-35](https://github.com/community/community/assets/1060056/be513f84-096d-4e3d-9165-9c67c4c53ae5)

Everything is flatten, so very hard to distinguish.

**Comment by fatmcgav** [comment]
Just to echo the above, reusable multi-job workflows in combination with the matrix strategy is very powerful, but the UI experience is extremely lacking...

**Comment by pauliusg** [comment]
The same for me, very hard to understand what is going on when you use matrix and reusable workflow:
![image](https://github.com/community/community/assets/1721818/ef1b13f2-a523-4000-97fe-665c83951b0f)

**Reply by Wineryeath** [reply]
bro suchas 
# .github/workflows/reusable-workflow.yml
name: Reusable Workflow with Explicit Dependencies

on:
  workflow_call:

jobs:
  setup:
    runs-on: ubuntu-latest
    outputs:
      setup-complete: ${{ steps.set-output.outputs.result }}
    steps:
      - name: Perform setup
        id: set-output
        run: echo "result=true" >> $GITHUB_OUTPUT
      
      - name: Log dependencies
        run: |
          echo "This is the SETUP job - no dependencies"
          echo "::notice::Downstream jobs will depend on this"

  build:
    needs: setup
    runs-on: ubuntu-latest
    steps:
      - name: Log dependencies
        run: |
          echo "This is the BUILD job - depends on SETUP"
          echo "::warning::Dependency: needs: setup"
          
      - name: Build artifacts
        run: echo "Building..."
huh huh

**Comment by bhidalgo-apolitical** [comment]
This is a very bad user experience, specially for monorepo structures which makes it almost impossible to understand what's going on.

Ideally each called workflow should be displayed in the UI individually - or at least have a toggle option to be able to see what has been generated from it. @vaind example clearly shows how a good design for this would look like.

Does someone from product has had the opportunity to have a look at this?

**Comment by snps-midhunn** [comment]
A flat list is really bad if the job names have similar keys. The fact that skipped jobs don't generate the name and display the generic job id is too painful. Too difficult to navigate and ton a lot of complains from devs.
![image](https://github.com/community/community/assets/107094461/058aad45-e0d3-4639-8ebe-8449436e57ae)

**Reply by bhidalgo-apolitical** [reply]
This would look way cleaner if the possibility to completely hide jobs was introduced too, instead of displaying it in the UI as skipped which simply clogs it with useless information.
GitLab does it very nicely, whenever a job does not comply with the `if` condition then it's skipped and hidden, so you only see what's actually running instead of displaying everything.

[a comment by an automated account is not shown]

**Comment by oliversaggau** [comment]
Any updates here?

**Comment by HugoSart** [comment]
Please someone fix this.

**Comment by branning** [comment]
Yes just discovered this limitation today, and it's got me reconsidering running this way. I think to work around this behavior you can use gh workflow run as the outer loop, instead of using reusable workflow.

**Comment by BiancaKatsina** [comment]
We ran into a **similar UI limitation with a matrix-based job setup**. We had a matrix of jobs, and each matrix execution included **multiple jobs with dependencies between them.**
The workflow itself behaved correctly, but **in the GitHub Actions UI the relationships between those jobs were not represented** clearly. The matrix executions appeared effectively flattened, so it was difficult to understand which jobs belonged together and how the dependencies between them were structured.
For more complex matrix/reusable workflow setups, it would be very helpful **if the UI preserved and visualized these relationships** instead of showing the jobs as a mostly flat list.
